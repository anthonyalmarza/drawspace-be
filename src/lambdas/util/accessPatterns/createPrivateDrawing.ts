import drawingEntity from 'lambdas/util/entities/drawing'
import { Drawing, UpdateReturnValues } from 'types'
import ksuid from 'ksuid'

export default async ({
    id = ksuid.randomSync().string,
    title,
    thumbnailUrl,
    drawStepsUrl,
    startTime,
    endTime,
    width,
    height,
    resolution,
    user,
}: {
    id?: string
    title: string
    thumbnailUrl: string
    drawStepsUrl: string
    startTime: string
    endTime: string
    width: number
    height: number
    resolution: number
    user: string
}): Promise<Drawing> => {
    const { Attributes: drawing } = await drawingEntity.update(
        {
            id,
            title,
            thumbnailUrl,
            drawStepsUrl,
            startTime,
            endTime,
            width,
            height,
            resolution,
            user,
        },
        {
            returnValues: UpdateReturnValues.ALL_NEW,
            conditions: { attr: 'pk', exists: false },
        }
    )
    return drawing
}
