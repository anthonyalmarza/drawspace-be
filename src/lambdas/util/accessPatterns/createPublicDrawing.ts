import drawingEntity from 'lambdas/util/entities/drawing'
import { Drawing, UpdateReturnValues } from 'types'
import ksuid from 'ksuid'
import dayjs from 'dayjs'

export default async ({
    id = ksuid.randomSync().string,
    title,
    thumbnailUrl,
    drawStepsUrl,
    startTime,
    endTime,
    width,
    height,
    resolution,
    user,
}: {
    id?: string
    title: string
    thumbnailUrl: string
    drawStepsUrl: string
    startTime: string
    endTime: string
    width: number
    height: number
    resolution: number
    user: string
}): Promise<Drawing> => {
    const dayjsNow = dayjs()
    const { Attributes: drawing } = await drawingEntity.update(
        {
            id,
            title,
            thumbnailUrl,
            drawStepsUrl,
            startTime,
            endTime,
            width,
            height,
            resolution,
            user,
            published: dayjsNow.toISOString(),
            gsi1pk: dayjsNow.format('YYYY-MM'),
            gsi1sk: dayjsNow.toISOString(),
        },
        {
            returnValues: UpdateReturnValues.ALL_NEW,
            conditions: { attr: 'pk', exists: false },
        }
    )
    return drawing
}
