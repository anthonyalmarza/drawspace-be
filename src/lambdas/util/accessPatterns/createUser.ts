import userEntity from 'lambdas/util/entities/user'
import { User, UpdateReturnValues } from 'types'

export default async ({
    userId,
    name,
}: {
    userId: string
    name: string
}): Promise<User> => {
    const { Attributes: user } = await userEntity.update(
        { id: userId, name },
        {
            returnValues: UpdateReturnValues.ALL_NEW,
            conditions: { attr: 'pk', exists: false },
        }
    )
    return user
}
