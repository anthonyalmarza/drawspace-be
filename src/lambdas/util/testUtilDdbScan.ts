import DynamoDB from 'aws-sdk/clients/dynamodb'

const { apiTableName } = require('lambdas/util/envVarsRestApi')
/* eslint-enable */
const DocumentClient = new DynamoDB.DocumentClient({
    endpoint: 'http://localhost:9001',
    accessKeyId: 'dynamo',
    secretAccessKey: 'devDummyKey',
    region: 'local',
})

export default () => DocumentClient.scan({ TableName: apiTableName }).promise()
