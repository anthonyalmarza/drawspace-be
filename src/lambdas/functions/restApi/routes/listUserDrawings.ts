import { API, HandlerFunction } from 'lambda-api'
import cognitoAuth from 'lambdas/functions/restApi/middleware/cognitoAuth'
import validateRequest from 'lambdas/functions/restApi/middleware/validateRequest'
import listUserDrawings from 'lambdas/util/accessPatterns/listUserDrawings'
import { parseStartKey } from 'lambdas/util/pagination'

const validateQueryParams = validateRequest(
    {
        startKey: { type: 'string' },
    },
    'query'
)

const validateUrlParams = validateRequest(
    {
        userId: { presence: true, type: 'string' },
    },
    'params'
)

const handler: HandlerFunction = async (req, res) => {
    const { startKey: b64StartKey } = req.query
    const { userId } = req.params
    const parsedStartKey = parseStartKey(b64StartKey)
    const result = await listUserDrawings({
        userId,
        startKey: parsedStartKey,
    })
    res.send(result)
}

export default (api: API): void => {
    api.get(
        '/users/:userId/drawings',
        cognitoAuth,
        validateUrlParams,
        validateQueryParams,
        handler
    )
}
