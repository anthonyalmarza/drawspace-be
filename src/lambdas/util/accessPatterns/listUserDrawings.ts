import drawingEntity from 'lambdas/util/entities/drawing'
import { Drawing, LastEvaluatedKey, DdbIndex } from 'types'
import { pagination } from 'lambdas/util/pagination'

type UserDrawings = {
    items: Drawing[]
    lastEvaluatedKey?: LastEvaluatedKey
}

export default async ({
    userId,
    startKey,
}: {
    userId: string
    startKey?: LastEvaluatedKey
}): Promise<UserDrawings> => {
    const {
        LastEvaluatedKey: lastEvaluatedKey,
        Items: drawings,
    } = await drawingEntity.query(`gsi2pkDrawing#${userId}`, {
        index: DdbIndex.GSI2,
        beginsWith: 'gsi2skDrawing#',
        reverse: true,
        ...(startKey ? { startKey } : {}),
    })
    return pagination<Drawing>({
        items: drawings,
        nextStartKey: lastEvaluatedKey,
    })
}
