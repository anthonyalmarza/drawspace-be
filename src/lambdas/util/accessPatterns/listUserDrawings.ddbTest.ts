import { advanceTo } from 'jest-date-mock'
import testUtilDdbScan from 'lambdas/util/testUtilDdbScan'
import createPrivateDrawing from 'lambdas/util/accessPatterns/createPrivateDrawing'
import createPublicDrawing from 'lambdas/util/accessPatterns/createPublicDrawing'
import createUser from 'lambdas/util/accessPatterns/createUser'
import listUserDrawings from 'lambdas/util/accessPatterns/listUserDrawings'

advanceTo(new Date(2020, 5, 29, 0, 0, 0))

const mockUserId = '1234123412341234'

test('listUserDrawings', async () => {
    const drawingMockCommon = {
        title: 'mockTitle',
        thumbnailUrl: 'mockThumbnailUrl',
        drawStepsUrl: 'mockDrawStepsUrl',
        startTime: 'mockStartTime',
        endTime: 'mockEndTime',
        width: 1000,
        height: 1000,
        resolution: 2,
        user: mockUserId,
    }
    await createUser({ userId: mockUserId, name: 'mock user name' })
    await Promise.all(
        Array.from({ length: 4 }, (x, i) =>
            Promise.all([
                createPrivateDrawing({
                    id: `mockId${i}Private`,
                    ...drawingMockCommon,
                }),
                createPublicDrawing({
                    id: `mockId${i}Public`,
                    ...drawingMockCommon,
                }),
            ])
        )
    )

    const res = await listUserDrawings(mockUserId)
    const scan = await testUtilDdbScan()
    expect(res).toMatchSnapshot()
    expect(scan).toMatchSnapshot()
})
