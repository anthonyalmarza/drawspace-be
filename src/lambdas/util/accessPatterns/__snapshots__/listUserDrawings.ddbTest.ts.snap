// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`listUserDrawings 1`] = `
Object {
  "items": Array [],
  "nextStartKey": undefined,
}
`;

exports[`listUserDrawings 2`] = `
Object {
  "Count": 9,
  "Items": Array [
    Object {
      "_ct": "2020-06-29T00:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-06-29T00:00:00.000Z",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi2pk": "gsi2pkDrawing#1234123412341234",
      "gsi2sk": "gsi2skDrawing#private#2020-06-29T00:00:00.000Z",
      "he": 1000,
      "pk": "drawing#mockId2Private",
      "r": 2,
      "sk": "drawing#mockId2Private",
      "st": "mockStartTime",
      "t": "mockTitle",
      "tu": "mockThumbnailUrl",
      "u": "1234123412341234",
      "wi": 1000,
    },
    Object {
      "_ct": "2020-06-29T00:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-06-29T00:00:00.000Z",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi2pk": "gsi2pkDrawing#1234123412341234",
      "gsi2sk": "gsi2skDrawing#private#2020-06-29T00:00:00.000Z",
      "he": 1000,
      "pk": "drawing#mockId3Private",
      "r": 2,
      "sk": "drawing#mockId3Private",
      "st": "mockStartTime",
      "t": "mockTitle",
      "tu": "mockThumbnailUrl",
      "u": "1234123412341234",
      "wi": 1000,
    },
    Object {
      "_ct": "2020-06-29T00:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-06-29T00:00:00.000Z",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi2pk": "gsi2pkDrawing#1234123412341234",
      "gsi2sk": "gsi2skDrawing#private#2020-06-29T00:00:00.000Z",
      "he": 1000,
      "pk": "drawing#mockId1Private",
      "r": 2,
      "sk": "drawing#mockId1Private",
      "st": "mockStartTime",
      "t": "mockTitle",
      "tu": "mockThumbnailUrl",
      "u": "1234123412341234",
      "wi": 1000,
    },
    Object {
      "_ct": "2020-06-29T00:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-06-29T00:00:00.000Z",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi2pk": "gsi2pkDrawing#1234123412341234",
      "gsi2sk": "gsi2skDrawing#private#2020-06-29T00:00:00.000Z",
      "he": 1000,
      "pk": "drawing#mockId0Private",
      "r": 2,
      "sk": "drawing#mockId0Private",
      "st": "mockStartTime",
      "t": "mockTitle",
      "tu": "mockThumbnailUrl",
      "u": "1234123412341234",
      "wi": 1000,
    },
    Object {
      "_ct": "2020-06-29T00:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-06-29T00:00:00.000Z",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi1pk": "gsi1pkDrawing#2020-06",
      "gsi1sk": "gsi1skDrawing#2020-06-29T00:00:00.000Z",
      "gsi2pk": "gsi2pkDrawing#1234123412341234",
      "gsi2sk": "gsi2skDrawing#public#2020-06-29T00:00:00.000Z",
      "he": 1000,
      "p": "2020-06-29T00:00:00.000Z",
      "pk": "drawing#mockId0Public",
      "r": 2,
      "sk": "drawing#mockId0Public",
      "st": "mockStartTime",
      "t": "mockTitle",
      "tu": "mockThumbnailUrl",
      "u": "1234123412341234",
      "wi": 1000,
    },
    Object {
      "_ct": "2020-06-29T00:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-06-29T00:00:00.000Z",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi1pk": "gsi1pkDrawing#2020-06",
      "gsi1sk": "gsi1skDrawing#2020-06-29T00:00:00.000Z",
      "gsi2pk": "gsi2pkDrawing#1234123412341234",
      "gsi2sk": "gsi2skDrawing#public#2020-06-29T00:00:00.000Z",
      "he": 1000,
      "p": "2020-06-29T00:00:00.000Z",
      "pk": "drawing#mockId3Public",
      "r": 2,
      "sk": "drawing#mockId3Public",
      "st": "mockStartTime",
      "t": "mockTitle",
      "tu": "mockThumbnailUrl",
      "u": "1234123412341234",
      "wi": 1000,
    },
    Object {
      "_ct": "2020-06-29T00:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-06-29T00:00:00.000Z",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi1pk": "gsi1pkDrawing#2020-06",
      "gsi1sk": "gsi1skDrawing#2020-06-29T00:00:00.000Z",
      "gsi2pk": "gsi2pkDrawing#1234123412341234",
      "gsi2sk": "gsi2skDrawing#public#2020-06-29T00:00:00.000Z",
      "he": 1000,
      "p": "2020-06-29T00:00:00.000Z",
      "pk": "drawing#mockId1Public",
      "r": 2,
      "sk": "drawing#mockId1Public",
      "st": "mockStartTime",
      "t": "mockTitle",
      "tu": "mockThumbnailUrl",
      "u": "1234123412341234",
      "wi": 1000,
    },
    Object {
      "_ct": "2020-06-29T00:00:00.000Z",
      "_et": "user",
      "_md": "2020-06-29T00:00:00.000Z",
      "dc": 0,
      "n": "mock user name",
      "pdc": 0,
      "pk": "user#1234123412341234",
      "sk": "user#1234123412341234",
    },
    Object {
      "_ct": "2020-06-29T00:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-06-29T00:00:00.000Z",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi1pk": "gsi1pkDrawing#2020-06",
      "gsi1sk": "gsi1skDrawing#2020-06-29T00:00:00.000Z",
      "gsi2pk": "gsi2pkDrawing#1234123412341234",
      "gsi2sk": "gsi2skDrawing#public#2020-06-29T00:00:00.000Z",
      "he": 1000,
      "p": "2020-06-29T00:00:00.000Z",
      "pk": "drawing#mockId2Public",
      "r": 2,
      "sk": "drawing#mockId2Public",
      "st": "mockStartTime",
      "t": "mockTitle",
      "tu": "mockThumbnailUrl",
      "u": "1234123412341234",
      "wi": 1000,
    },
  ],
  "ScannedCount": 9,
}
`;
